qiita_db/util.py

```diff
@@ -2311,3 +2317,476 @@ def send_email(to, subject, body):
+                else:
+                    right = middle
+
+            # proceed with binary search in a window 10k to the right
+            if left >= right and cnt < depth:
+                sl += 10000
+                sr += 10000
+                left = sl
+                right = sr
+
+            cnt += 1
+
+        # it's ok if we have a couple failures
+        failure_tolerance = 3
+        is_acceptable_based_on_failures = (
+            prev_failures <= failure_tolerance or abs(
+                prev_failures - best_failures) < failure_tolerance or
+            best_failures == np.inf)
+
+        # case where less failures
+        if is_acceptable_based_on_failures:
+            if min_max <= best_max:
+                best_failures = prev_failures
+                best_max = min_max
+                best_model = model
+                best_result = res
+    return best_model, best_result
+
+
+def _resource_allocation_custom_loss(params, x, y, model, p):
+    """Helper function for resource allocation plot. Calculates custom loss
+    for given model.
+
+    Parameters
+    ----------
+    params: list, required
+        Initial list of integers for the given model
+    x: pandas.Series (pandas column), required
+        Represents x data for the function calculation
+    y: pandas.Series (pandas column), required
+        Represents y data for the function calculation
+    models: list, required
+        List of functions that will be used for visualization
+    p: int, required
+        Penalty weight for custom loss function
+
+    Returns
+    ----------
+    float
+        The mean of the list returned by the loss calculation (np.where)
+    """
+    k, a, b = params
+
+    residuals = y - model(x, k, a, b)
+    # Apply a heavier penalty to points below the curve
+    penalty = p
+    weighted_residuals = np.where(residuals > 0, penalty * residuals**2,
+                                  residuals**2)
+    return np.mean(weighted_residuals)
+
+
+def _resource_allocation_failures(_df, k, a, b, model, COL_NAME, type_):
+    """Helper function for resource allocation plot. Creates a dataframe with
+    failures.
+
+    Parameters
+    ----------
+    _df: pandas.Dataframe, required
+        Represents dataframe containing current jobs data
+    k: int, required
+        k constant in a model
+    a: int, required
+        a constant in a model
+    b: int, required
+        b constant in a model
+    model: function, required
+        Current function
+    COL_NAME: str, required
+        Specifies x axis for the graph
+    type_: str, required
+        Specifies for which type we're getting failures (e.g. MaxRSSRaw)
+
+    Returns
+    ----------
+    pandas.Dataframe
+        Dataframe containing failures for current type.
+    """
+
+    x_plot = np.array(_df[COL_NAME])
+    _df[f'c{type_}'] = model(x_plot, k, a, b)
+    failures_df = _df[_df[type_] > _df[f'c{type_}']]
+    return failures_df
+
+
+def mem_model1(x, k, a, b):
```

Comment:
It seems odd to implement these different memory and time models as one line functions. Why incur the overhead of a function call, when these can simply be defined as consts within the one function that uses them?